src/Unsafe/Linear.hs

```diff
@@ -25,37 +32,125 @@ module Unsafe.Linear
     toLinear,
     toLinear2,
     toLinear3,
+    ToLinearN (..),
+    toLinearN
   )
   where
 
-import qualified Unsafe.Coerce as NonLinear
-import GHC.Exts (TYPE, RuntimeRep)
+import Data.Kind (Type, Constraint)
+import Unsafe.Coerce (UnsafeEquality (..), unsafeEqualityProof)
+import GHC.Exts (TYPE, RuntimeRep (..))
+import GHC.Types (Multiplicity (..))
+import GHC.TypeNats
+import GHC.TypeLits (TypeError, ErrorMessage (..))
 
 -- | Linearly typed @unsafeCoerce@
 coerce :: forall a b. a %1-> b
 coerce a =
-  case NonLinear.unsafeEqualityProof @a @b of
-    NonLinear.UnsafeRefl -> a
+  case unsafeEqualityProof @a @b of
+    UnsafeRefl -> a
 {-# INLINE coerce #-}
 
+-- TODO: In GHC 9.0, using toLinearN to implement toLinear2 and toLinear3
+-- produces larger Core, which presumably could hamper inlining and such.
+-- I suspect this may have been fixed in GHC 9.2, but I haven't checked.
+-- If so, we might want to actually define all the functions using
+-- toLinearN.
+
 -- | Converts an unrestricted function into a linear function
 toLinear
   :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep)
      (a :: TYPE r1) (b :: TYPE r2) p.
      (a %p-> b) %1-> (a %1-> b)
-toLinear = coerce
+toLinear f = case unsafeEqualityProof @p @'One of
+  UnsafeRefl -> f
 
 -- | Like 'toLinear' but for two-argument functions
 toLinear2
   :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep) (r3 :: RuntimeRep)
      (a :: TYPE r1) (b :: TYPE r2) (c :: TYPE r3) p q.
      (a %p-> b %q-> c) %1-> (a %1-> b %1-> c)
-toLinear2 = coerce
+toLinear2 f = case unsafeEqualityProof @'(p,q) @'( 'One, 'One) of
+  UnsafeRefl -> f
 
 -- | Like 'toLinear' but for three-argument functions
 toLinear3
   :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep)
      (r3 :: RuntimeRep) (r4 :: RuntimeRep)
      (a :: TYPE r1) (b :: TYPE r2) (c :: TYPE r3) (d :: TYPE r4) p q r.
      (a %p-> b %q-> c %r-> d) %1-> (a %1-> b %1-> c %1-> d)
-toLinear3 = coerce
+toLinear3 f = case unsafeEqualityProof @'(p,q,r) @'( 'One, 'One, 'One) of
+  UnsafeRefl -> f
+
+
+-- | @toLinearN@ subsumes the functionality of 'toLinear1', 'toLinear2', and
+-- 'toLinear3'. In particular, @toLinearN \@n@ unsafely changes the
+-- multiplicities of the first @n@ arrows from any multiplicity to any
+-- other multiplicity.
+--
+-- == Examples
+--
+-- @
+-- toLinearN \@2 :: (a -> b -> Int) -> a %1-> b %1-> Int
+-- 'toLinear3' = toLinearN \@3
+-- @
+toLinearN :: forall n f g m. ToLinearN n f g => f %m-> g
```

Comment:
I don't know the conventions; is it better to use `%m` here or `%1`?

Reply:
We tend to use `%m` for functions which are meant to be compatible with an existing library (typically base; things like `const`) and %1 otherwise. Especially for a function like this which we never intend to pass as an argument for a function.

Reply:
Okay, I guess I'll make them unconditionally linear.